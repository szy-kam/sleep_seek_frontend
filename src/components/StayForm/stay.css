.message {
    position: fixed;
    top: 20%;
    left: 50%;
    background-color: blue;
    padding: 20px;
    color: white;
    border-radius: 5px;
    z-index: 4;
}

.addStayCompoment, .editStayCompoment{
    display: flex;
    flex-direction: column;
    align-items: center;
}
