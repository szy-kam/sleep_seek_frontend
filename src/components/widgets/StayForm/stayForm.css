/* *********************************************
******************   Edit Form **************
********************************************* */
.stayEditCompoment {
    display: flex;
    align-items: center;
    flex-direction: column;
}

.editForm {
    margin: 20px auto;
    padding: 30px;
    border: 1px solid black;
    display: flex;
    flex-direction: column;
}

.editForm input,
.editForm textarea {
    border: 1px solid black;
    margin: 10px;
    padding: 6px;
}

.editForm textarea {
    height: 100px;
    width: 400px;
}

.message {
    position: fixed;
    top: 20%;
    left: 50%;
    background-color: blue;
    padding: 20px;
    color: white;
    border-radius: 5px;
}

.thumbs {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
}

.thumbs img {
    width: auto;
    height: auto;
    max-width: 150px;
    margin: 10px;
}

.selectedPhoto {
    border: 4px solid red;
}
