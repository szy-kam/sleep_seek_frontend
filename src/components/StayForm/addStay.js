import React, { Component } from "react";
import { AddStayRepository } from "../../repository/stay";
import StayForm from "./stayForm";
import style from "./stay.css";
import { withTranslation } from "react-i18next";
import PopupComponent from "../widgets/PopupComponent/popupComponent";
import LoadingComponent from "../widgets/LoadingComponent/loadingComponent";

class AddStay extends Component {

    state = {
        message: "",
        showPopup: false,
    }

    handleSubmit = (stay, files) => {
        this.setState({ showPopup: true, message: "ADDING_STAY" })
        const { t } = this.props;
        AddStayRepository(stay, files)
            .then((response) => {
                if (response.ok) {
                    response.json()
                        .then(data => {
                            this.setState({ message: t("STAY_ADDED"), showPopup: true })
                            const editUrl = `/stays/editAccommodations/${data}`
                            this.redirectUser(editUrl)
                        })
                }
                else {
                    response.json().then(err => {
                        console.log(err);
                    })
                    this.setState({ message: t(`ERROR_${response.status}`) })
                }
            });
    };

    redirectUser = (url = "/my-account") => {
        this.props.history.push(url);
    }


    render() {
        const { t } = this.props;
        return (
            <div className={style.addStayCompoment}>
                {this.state.showPopup && <PopupComponent >
                    <h3>{t(this.state.message)}</h3>
                    <LoadingComponent />
                </PopupComponent>}
                <StayForm
                    handleSubmit={this.handleSubmit}
                    redirectUser={this.redirectUser}
                />
            </div>
        );
    }
}
export default withTranslation()(AddStay);
