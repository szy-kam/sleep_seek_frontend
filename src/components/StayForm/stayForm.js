import React, { Component } from "react";
import FileUploader from "../widgets/FileUploader/fileUploader";
import style from "./stayForm.css";
import { GetStayByIdRepository, GetAllStayCategories, DeleteStayPhotoRepository } from "../../repository/stay";
import { withTranslation } from "react-i18next";
import { STAY } from "../../config";
import StayMap from "../widgets/StayMap/stayMap";
import { OpenStreetMapProvider } from "leaflet-geosearch";
import countrysList from "../../repository/countrysList";
import PropertiesForm from "../PropertiesForm/propertiesForm";
import { connect } from "react-redux";

class StayForm extends Component {
    state = {
        stay: STAY,
        newPhotos: [],
        categories: []
    };

    componentDidMount() {
        if (this.props.getStay) {
            GetStayByIdRepository(this.props.getStay)
                .then((response) => {
                    if (response.ok) {
                        return response.json()
                    }
                    else {
                        this.props.redirectUser("/my-account")
                    }
                })
                .then(data => {
                    if (this.props.user && this.props.user.username === data.username) {
                        this.setState({ stay: data })
                    }
                    else {
                        this.props.redirectUser("/my-account")
                    }

                });
        }
        else {
            // To set default values in selects
            const stay = this.state.stay
            stay.address.country = "Polska"
            stay.category = "HOTEL"
            this.setState({
                stay: stay
            })
        }

        GetAllStayCategories()
            .then(response => response.json())
            .then(data => {
                if (Array.isArray(data)) {
                    this.setState({ categories: data })
                }
            })
    }

    handleInput = (event, field) => {
        const newStay = {
            ...this.state.stay,
        };

        if (["city", "street", "zipCode", "country"].indexOf(field) >= 0)
            newStay.address[field] = event.target.value;
        else newStay[field] = event.target.value;

        this.setState({
            stay: newStay,
        });
    };

    mapReposition = () => {
        const address =
            this.state.stay.address.street +
            " " +
            this.state.stay.address.city +
            " " +
            this.state.stay.address.zipCode +
            " " +
            this.state.stay.address.country;

        if (address.length > 5) {
            const provider = new OpenStreetMapProvider();
            provider.search({ query: address }).then((result) => {
                if (result.length >= 1) {
                    const newStay = {
                        ...this.state.stay,
                    };
                    newStay.address.longitude = result[0].x;
                    newStay.address.latitude = result[0].y;
                    this.setState({
                        stay: newStay,
                    });
                } else {
                    const newStay = {
                        ...this.state.stay,
                    };
                    newStay.address.longitude = null;
                    newStay.address.latitude = null;
                    this.setState({
                        stay: newStay,
                    });
                }
            });
        }
    };

    countrysOptions = () => {
        return countrysList.map((item, i) => {
            return <option key={i}>{item}</option>;
        });
    };

    stayCategoryOptions = () => {
        return this.state.categories.map((item, i) => {
            return <option key={i} value={item}>{this.props.t(item)}</option>;
        });

    }

    handleSubmit = (event) => {
        event.preventDefault();
        this.props.handleSubmit(this.state.stay, this.state.newPhotos, this.state.properties);
    };

    onDrop = (acceptedFiles) => {
        acceptedFiles.map((file) =>
            Object.assign(file, {
                preview: URL.createObjectURL(file),
            })
        );

        this.setState({
            newPhotos: [...this.state.newPhotos, ...acceptedFiles]
        });
    };

    mainPhotoSelected = (e) => {
        const newStay = {
            ...this.state.stay,
        };
        newStay.mainPhoto = e.target.alt;
        this.setState({
            stay: newStay,
        });
    };

    thumbs = () => {
        if (this.state.newPhotos && this.state.newPhotos.length > 0)
            return this.state.newPhotos.map((file, i) => (
                <div className={style.photoContainer} key={i}>
                    <div id={file.name} className={style.deletePhoto} onClick={this.handleUndoUpload}>x</div>
                    <img
                        src={file.preview}
                        key={file.name}
                        alt={file.name}
                        onClick={this.mainPhotoSelected}
                        className={file.name === this.state.stay.mainPhoto ? style.selectedPhoto : null}
                    />
                </div>
            ));
        else return null;
    };

    handlePropertiesInput = (val) => {
        const newStay = {
            ...this.state.stay,
        };
        newStay.properties = val

        this.setState({
            stay: newStay
        });
    }

    handleDeletePhoto = (event) => {
        let newStay = {
            ...this.state.stay,
        };

        newStay.photos = newStay.photos.filter(photo => {
            return photo !== event.target.id
        });

        if (event.target.id === this.state.stay.mainPhoto) {
            newStay.mainPhoto = ""
        }

        this.setState({
            stay: newStay,
        });

        DeleteStayPhotoRepository(event.target.id)
            .catch(err => console.log(err))
    }

    handleUndoUpload = (event) => {
        let photos = this.state.newPhotos
        photos = photos.filter(photo => {
            return photo.name !== event.target.id
        });
        if (event.target.id === this.state.stay.mainPhoto) {
            let newStay = { ...this.state.stay }
            newStay.mainPhoto = ""
            this.setState({
                newPhotos: photos,
                stay: newStay
            });
        } else {
            this.setState({
                newPhotos: photos,
            });
        }

    }

    renderPhotos = () => {
        if (this.state.stay.photos) {
            return this.state.stay.photos.map((item, i) => {
                return (
                    <div className={style.photoContainer} key={i}>
                        <div id={item} className={style.deletePhoto} onClick={this.handleDeletePhoto}>x</div>
                        <div className={style.deletedPhoto}></div>
                        <img
                            src={item}
                            key={i}
                            alt={item}
                            onClick={this.mainPhotoSelected}
                            className={item === this.state.stay.mainPhoto ? style.selectedPhoto : null}
                        />
                    </div>
                )
            })
        }
    }


    render() {
        const { t } = this.props;
        return (
            <div className={style.stayEditCompoment}>
                <form onSubmit={this.handleSubmit} className={style.editForm}>
                    <label>{t("NAME")}</label>
                    <input
                        onChange={(event) => this.handleInput(event, "name")}
                        value={this.state.stay.name}
                        maxLength="100"
                        required
                    />
                    <label>{t("CATEGORY")}</label>
                    <select
                        onChange={(event) => this.handleInput(event, "category")}
                        value={this.state.stay.category}
                    >
                        {this.stayCategoryOptions()}
                    </select>
                    <label>{t("COUNTRY")}</label>
                    <select
                        onChange={(event) => this.handleInput(event, "country")}
                        value={this.state.stay.address.country}
                    >
                        {this.countrysOptions()}
                    </select>
                    <label>{t("CITY")}</label>
                    <input
                        onChange={(event) => this.handleInput(event, "city")}
                        onBlur={this.mapReposition}
                        value={this.state.stay.address.city}
                        maxLength="100"
                        required
                    />
                    <label>{t("STREET")}</label>
                    <input
                        onChange={(event) => this.handleInput(event, "street")}
                        onBlur={this.mapReposition}
                        value={this.state.stay.address.street}
                        maxLength="100"
                        required
                    />
                    <label>{t("ZIP_CODE")}</label>
                    <input
                        onChange={(event) => this.handleInput(event, "zipCode")}
                        onBlur={this.mapReposition}
                        value={this.state.stay.address.zipCode}
                        placeholder="00-000"
                        maxLength="6"
                    />
                    {this.state.stay.address.latitude && (
                        <div className={style.map}>
                            <StayMap
                                position={[
                                    this.state.stay.address.latitude,
                                    this.state.stay.address.longitude,
                                ]}
                                zoom={16}
                            />
                        </div>

                    )}
                    <label>{t("MIN_PRICE")} ({t("CURRENCY_SYMBOL")})</label>
                    <input
                        onChange={(event) => this.handleInput(event, "minPrice")}
                        value={this.state.stay.minPrice}
                        placeholder={t("MIN_PRICE_INFO")}
                        type="number"
                        required
                        min="1"
                    />
                    <label>{t("PHONE_NUMBER")}</label>
                    <input
                        onChange={(event) => this.handleInput(event, "phoneNumber")}
                        value={this.state.stay.phoneNumber}
                        maxLength="15"
                    />
                    <label>{t("EMAIL")}</label>
                    <input
                        onChange={(event) => this.handleInput(event, "email")}
                        value={this.state.stay.email}
                        type="email"
                        maxLength="50"
                    />
                    <label>{t("DESCRIPTION")}</label>
                    <textarea
                        onChange={(event) => this.handleInput(event, "description")}
                        value={this.state.stay.description}
                        maxLength="10000"
                    />
                    <label>{t("PHOTOS")}</label>
                    <div className={style.photos}>{this.renderPhotos()}</div>
                    <FileUploader onDrop={this.onDrop} files={this.state.stay.photos} />
                    <div className={style.thumbs}>{this.thumbs()}</div>

                    {!this.state.stay.mainPhoto && (this.state.newPhotos.length || this.state.stay.photos.length) ? <div className={style.selectMainPhoto}>{t("SELECT_MAIN_PHOTO")}</div> : null}

                    <label>{t("PROPERTIES")}</label>
                    <div className={style.properties}>
                        <PropertiesForm handleInput={this.handlePropertiesInput} stay={true} properties={this.state.stay.properties} />
                    </div>

                    <button type="submit" className={style.submitButton}>
                        {this.props.getStay ? t("SAVE") : t("ADD_STAY")}
                    </button>

                    {this.props.handleDelete ? (
                        <button onClick={this.props.handleDelete} className={style.deleteButton}>{t("DELETE_STAY")}</button>
                    ) : null}

                </form>
            </div>
        );
    }
}

const mapStateToProps = state => ({
    user: state.user.user
})

export default connect(mapStateToProps)(withTranslation()(StayForm))