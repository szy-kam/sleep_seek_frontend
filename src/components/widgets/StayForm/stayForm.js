import React, { Component } from "react";
import FileUploader from "../FileUploader/fileUploader";
import style from "./stayForm.css";
import { GetStayByIdRepository } from "../../../repository/stay";
import { withTranslation } from "react-i18next";
import { STAY } from "../../../config";
import StayMap from "../StayMap/stayMap";
import { OpenStreetMapProvider } from "leaflet-geosearch";
import countrysList from "../../../repository/countrysList";

class StayForm extends Component {
    state = {
        stay: STAY,
    };

    componentDidMount() {
        if (this.props.getStay)
            GetStayByIdRepository(this.props.getStay).then((response) =>
                this.setState({ stay: response })
            );
    }

    handleInput = (event, field) => {
        const newStay = {
            ...this.state.stay,
        };

        if (["city", "street", "zipCode", "country"].indexOf(field) >= 0)
            newStay.address[field] = event.target.value;
        else newStay[field] = event.target.value;

        this.setState({
            stay: newStay,
        });
    };

    mapReposition = () => {
        const address =
            this.state.stay.address.street +
            " " +
            this.state.stay.address.city +
            " " +
            this.state.stay.address.zipCode +
            " " +
            this.state.stay.address.country;

        if (address.length > 5) {
            const provider = new OpenStreetMapProvider();
            provider.search({ query: address }).then((result) => {
                if (result.length >= 1) {
                    const newStay = {
                        ...this.state.stay,
                    };
                    newStay.address.longitude = result[0].x;
                    newStay.address.latitude = result[0].y;
                    this.setState({
                        stay: newStay,
                    });
                } else {
                    const newStay = {
                        ...this.state.stay,
                    };
                    newStay.address.longitude = null;
                    newStay.address.latitude = null;
                    this.setState({
                        stay: newStay,
                    });
                }
            });
        }
    };

    countrysOptions = () => {
        return countrysList.map((item, i) => {
            return <option key={i}>{item}</option>;
        });
    };

    handleSubmit = (event) => {
        event.preventDefault();
        this.props.handleSubmit(this.state.stay, this.state.newPhotos);
    };

    onDrop = (acceptedFiles) => {

        acceptedFiles.map((file) =>
            Object.assign(file, {
                preview: URL.createObjectURL(file),
            })
        );

        this.setState({
            newPhotos: acceptedFiles,
        });
    };

    mainPhotoSelected = (e) => {
        const newStay = {
            ...this.state.stay,
        };
        newStay.mainPhoto = e.target.alt;
        this.setState({
            stay: newStay,
        });
    };

    thumbs = () => {
        if (this.state.newPhotos && this.state.newPhotos.length > 0)
            return this.state.newPhotos.map((file) => (
                <img
                    src={file.preview}
                    key={file.name}
                    alt={file.name}
                    onClick={this.mainPhotoSelected}
                    className={file.name === this.state.stay.mainPhoto ? style.selectedPhoto : null}
                />
            ));
        else return null;
    };

    render() {
        const { t } = this.props;

        return (
            <div className={style.stayEditCompoment}>
                <form onSubmit={this.handleSubmit} className={style.editForm}>
                    <label>{t("NAME")}</label>
                    <input
                        onChange={(event) => this.handleInput(event, "name")}
                        value={this.state.stay.name}
                    />
                    <label>{t("COUNTRY")}</label>
                    <select
                        onChange={(event) => this.handleInput(event, "country")}
                        value={this.state.stay.address.country}
                    >
                        {this.countrysOptions()}
                    </select>
                    <label>{t("CITY")}</label>
                    <input
                        onChange={(event) => this.handleInput(event, "city")}
                        onBlur={this.mapReposition}
                        value={this.state.stay.address.city}
                    />
                    <label>{t("STREET")}</label>
                    <input
                        onChange={(event) => this.handleInput(event, "street")}
                        onBlur={this.mapReposition}
                        value={this.state.stay.address.street}
                    />
                    <label>{t("ZIP_CODE")}</label>
                    <input
                        onChange={(event) => this.handleInput(event, "zipCode")}
                        onBlur={this.mapReposition}
                        value={this.state.stay.address.zipCode}
                    />
                    {this.state.stay.address.latitude && (
                        <StayMap
                            position={[
                                this.state.stay.address.latitude,
                                this.state.stay.address.longitude,
                            ]}
                            zoom={16}
                        />
                    )}
                    <label>{t("MIN_PRICE")}</label>
                    <input
                        onChange={(event) => this.handleInput(event, "minPrice")}
                        value={this.state.stay.price}
                    />
                    <label>{t("PHONE_NUMBER")}</label>
                    <input
                        onChange={(event) => this.handleInput(event, "phoneNumber")}
                        value={this.state.stay.contactInfo}
                    />
                    <label>{t("EMAIL")}</label>
                    <input
                        onChange={(event) => this.handleInput(event, "email")}
                        value={this.state.stay.contactInfo}
                    />
                    <label>{t("DESCRIPTION")}</label>
                    <textarea
                        onChange={(event) => this.handleInput(event, "description")}
                        value={this.state.stay.description}
                    />
                    <label>{t("PHOTOS")}</label>
                    <FileUploader onDrop={this.onDrop} files={this.state.stay.photos} />
                    <div className={style.thumbs}>{this.thumbs()}</div>

                    {!this.state.stay.mainPhoto ? t("SELECT_MAIN_PHOTO") : null}

                    <button type="submit">
                        {this.props.getStay ? t("EDIT_STAY") : t("ADD_STAY")}
                    </button>

                    {this.props.handleDelete ? (
                        <button onClick={this.props.handleDelete}>{t("DELETE_STAY")}</button>
                    ) : null}
                    
                </form>
            </div>
        );
    }
}
export default withTranslation()(StayForm);
