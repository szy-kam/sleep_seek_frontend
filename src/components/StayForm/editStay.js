import React, { Component } from "react";
import {
    EditStayRepository,
    DeleteStayByIdRepository,
} from "../../repository/stay";
import StayForm from "./stayForm";
import style from "./stay.css";
import { withTranslation } from "react-i18next";
import { Link } from "react-router-dom";
import LoadingComponent from "../widgets/LoadingComponent/loadingComponent";
import PopupComponent from "../widgets/PopupComponent/popupComponent";

class EditStay extends Component {
    state = {
        message: "",
        showPopup: false
    };

    handleSubmit = (stay, files) => {
        this.setState({ showPopup: true, message: "SAVING_STAY" })
        const { t } = this.props;
        EditStayRepository(stay, files)
            .then((response) => {
                if (response.ok) {
                    this.redirectUser()
                }
                else {
                    console.log(response);
                    this.setState({ message: t(`ERROR_${response.status}`) })
                }
            })
    };

    handleDelete = (e) => {
        e.preventDefault();
        DeleteStayByIdRepository(this.props.match.params.id)
            .then(() =>
                this.redirectUser()
            );
    };

    redirectUser = (url = "/my-account") => {
        this.props.history.push(url);
    };

    render() {
        const { t } = this.props;
        return (
            <div className={style.editStayCompoment}>
                {this.state.showPopup && <PopupComponent >
                    <h3>{t(this.state.message)}</h3>
                    <LoadingComponent />
                </PopupComponent>}
                <StayForm
                    handleInput={this.handleInput}
                    stay={this.state.stay}
                    handleSubmit={this.handleSubmit}
                    handleDelete={this.handleDelete}
                    getStay={this.props.match.params.id}
                    redirectUser={this.redirectUser}
                />
                <Link to={`/stays/editAccommodations/${this.props.match.params.id}`}><button>{t('EDIT_ACCOMMODATIONS')}</button></Link>
            </div>
        );
    }
}
export default withTranslation()(EditStay);
