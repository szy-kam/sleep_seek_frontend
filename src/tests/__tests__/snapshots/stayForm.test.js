import React from 'react'
import StayForm from '../../../components/StayForm/stayForm'
import { render } from 'enzyme';
import { stayMock1 } from '../../mocks'
import { BrowserRouter as Router } from 'react-router-dom';

it("Snapshot test. With stay", () => {
    const component = render(<Router><StayForm stay={stayMock1} /></Router>)
    expect(component).toMatchSnapshot()
})

