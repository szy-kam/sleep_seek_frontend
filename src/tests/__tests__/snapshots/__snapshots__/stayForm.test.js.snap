// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test. Default template. No stay 1`] = `
initialize {
  "0": Node {
    "attribs": Object {},
    "children": Array [],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
